packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.swift

```diff
@@ -0,0 +1,422 @@
+  public convenience init(
+    receiptManager: FIAPReceiptManager,
+    handlerFactory: @escaping (SKRequest) -> FIAPRequestHandler
+  ) {
+    self.init(receiptManager: receiptManager)
+    self.handlerFactory = handlerFactory
+  }
+
+  // This is the designated initializer
+  public convenience init(registrar: FlutterPluginRegistrar) {
+    self.init(receiptManager: FIAPReceiptManager())
+    self.registrar = registrar
+
+    self.paymentQueueHandler = FIAPaymentQueueHandler(
+      queue: SKPaymentQueue.default(),
+      transactionsUpdated: { [weak self] transactions in
+        self?.handleTransactionsUpdated(transactions)
+      },
+      transactionRemoved: { [weak self] transactions in
+        self?.handleTransactionsRemoved(transactions)
+      },
+      restoreTransactionFailed: { [weak self] error in
+        self?.handleTransactionRestoreFailed(error as NSError)
+      },
+      restoreCompletedTransactionsFinished: { [weak self] () -> Void in
+        self?.restoreCompletedTransactionsFinished()
+      },
+      shouldAddStorePayment: { [weak self] (payment: SKPayment, product: SKProduct) -> Bool in
+        return self?.shouldAddStorePayment(payment: payment, product: product) ?? false
+      },
+      updatedDownloads: { [weak self] _ in
+        self?.updatedDownloads()
+      },
+      transactionCache: FIATransactionCache())
+    #if os(iOS)
+      let messenger = registrar.messenger()
+    #endif
+    #if os(macOS)
+      let messenger = registrar.messenger
+    #endif
+    transactionObserverCallbackChannel = FlutterMethodChannel(
+      name: "plugins.flutter.io/in_app_purchase", binaryMessenger: messenger)
+  }
+
+  public func canMakePaymentsWithError(_ error: AutoreleasingUnsafeMutablePointer<FlutterError?>)
+    -> NSNumber?
+  {
+    return SKPaymentQueue.canMakePayments() as NSNumber
+  }
+
+  public func transactionsWithError(_ error: AutoreleasingUnsafeMutablePointer<FlutterError?>)
+    -> [SKPaymentTransactionMessage]?
+  {
+
+    let transactions = getPaymentQueueHandler().getUnfinishedTransactions()
+    return transactions.compactMap {
+      FIAObjectTranslator.convertTransaction(toPigeon: $0)
+    }
+  }
+
+  public func storefrontWithError(_ error: AutoreleasingUnsafeMutablePointer<FlutterError?>)
+    -> SKStorefrontMessage?
+  {
+    if #available(iOS 13.0, *), let storefront = getPaymentQueueHandler().storefront {
+      return FIAObjectTranslator.convertStorefront(toPigeon: storefront)
+    }
+    return nil
+  }
+
+  public func startProductRequestProductIdentifiers(
+    _ productIdentifiers: [String],
+    completion: @escaping (SKProductsResponseMessage?, FlutterError?) -> Void
+  ) {
+    let request = getProductRequest(withIdentifiers: Set(productIdentifiers))
+    let handler = handlerFactory(request)
+    requestHandlers.insert(handler)
+
+    handler.startProductRequest { [weak self] response, startProductRequestError in
```

Comment:
this is a pretty long closure. can you check `weak self` first as the first line? In case of some race condition during the execution of this block

Reply:
Like check if `weak self` exists?

Reply:
you can do `guard let self = else else { return }`